src/metabase/search/ingestion.clj

```diff
@@ -115,25 +115,43 @@
   false)
 
 (defn update!
-  "Update all active engines' existing indexes with the given documents"
-  [documents-reducible]
+  "Update all active engines' existing indexes with the given documents. Passed remove-documents will be deleted from the index."
+  [documents-reducible remove-documents]
   (when-let [engines (seq (search.engine/active-engines))]
     (if (= 1 (count engines))
-      (search.engine/update! (first engines) documents-reducible)
-      ;; TODO um, multiplexing over the reducible awkwardly feels strange. We at least use a magic number for now.
-      (doseq [batch (eduction (partition-all 150) documents-reducible)
-              e     engines]
-        (search.engine/update! e batch)))))
+      (u/prog1 (search.engine/update! (first engines) documents-reducible)
+        (doseq [[group ids] (u/group-by first second remove-documents)]
+          (search.engine/delete! (first engines) group ids)))
+      (u/prog1
+        ;; TODO um, multiplexing over the reducible awkwardly feels strange. We at least use a magic number for now.
+        (doseq [batch (eduction (partition-all 150) documents-reducible)
+                e engines]
+          (search.engine/update! e batch))
+        (doseq [[group ids] (u/group-by first second remove-documents)]
+          (search.engine/delete! (first engines) group ids))))))
```

Comment:
1. There's a bug here the you're not deleting from all the engines in the second if branch.
2. You're doing the same stuff in both the prog1 bodies, so you could just put the prog1 around the if instead?

(2) doesn't apply if you maybe `reducible-engines` a reducible though. I prefer making it a reducible, so this point is just for if you push back on that 😄